src/cid.js

```diff
@@ -3,48 +3,310 @@ import * as Digest from './hashes/digest.js'
 import { base58btc } from './bases/base58.js'
 import { base32 } from './bases/base32.js'
 import { coerce } from './bytes.js'
+// Linter can see that API is used in types.
+// eslint-disable-next-line
+import * as API from './cid/interface.js'
+
+// This way TS will also expose all the types from module
+export * from './cid/interface.js'
+
+/**
+ * @template {number} Format
+ * @template {number} Alg
+ * @template {API.CIDVersion} Version
+ * @template T
+ * @param {API.CID<Format, Alg, Version>|T} input
+ * @returns {API.CID<Format, Alg, Version>|null}
+ */
+export const asCID = (input) => {
```

Comment:
pulling all of these up increases the number of things exported from this module, is there a good reason for this? we've seen problems being surprisingly breaking for types when we've done this in the past, I can't see an obvious path to that for this module but maybe we shouldn't be moving things around so much in this PR if they are not necessary for it? Or is there some interface pattern you want to take advantage of that avoids touching the class directly?

Reply:
I want to stop using class in all the code we're writing and write things against interface. Changes here make this possible by making functions in both input and output refer to interface while class continues to stick to the class and generally behave as it used to.

This also can't break existing code because these functions did not existed before. If moving these into separate file would make it less scary I'm happy to do that too.

Reply:
oh, that's pretty nice; I guess the downside of having them in the same file is that when you `import * as CID from ...` you get `CID.CID` as legacy, but maybe that's not so bad.